File: elixir/lib/parser.ex
require OK

# {:error,nil}|{:ok,{T,str}}
defmodule Parser do
  defp expect(<<x::bytes-size(1)>><>xs,f) do
    if f.(x) do
      {:ok,{x,xs}}
    else
      {:error,nil}
    end
  end

  defp expect(_,_),do: {:error,nil}

  defp char(s,c),do: expect(s,fn x -> x==c end)

  defp number("-"<>s),do: number("",-1,s)
  defp number(s),do: number("",1,s)

  defp number(n,g,<<c::bytes-size(1)>><>s) when c=="0" or c=="1" or c=="2" or c=="3" or c=="4" or c=="5" or c=="6" or c=="7" or c=="8" or c=="9",do: number(n<>c,g,s)
  defp number("0"<><<_::bytes-size(1)>><>_,_,_),do: {:error, nil}
  defp number(n,g,s) do
    case Integer.parse(n) do
      :error -> {:error, nil}
      {x,_} -> {:ok, {Rational.from_int(x*g),s}}
    end
  end

  defp eof(""),do: {:ok,{nil,""}}
  defp eof(_),do: {:error,nil}

  defp expr(s) do
    OK.for do
      {x,s}<-term(s)
      res<-expr(s,x)
    after
      res
    end
  end
  defp expr("+"<>s,n) do
    OK.for do
      {x,s}<-term(s)
      res<-expr(s,Rational.add(n,x))
    after
      res
    end
  end
  defp expr("-"<>s,n) do
    OK.for do
      {x,s}<-term(s)
      res<-expr(s,Rational.sub(n,x))
    after
      res
    end
  end
  defp expr(s,n) do
    {:ok,{n,s}}
  end
  def parse(s) do
    OK.for do
      {x,s}<-expr(s)
      _<-eof(s)
    after
      x
    end
  end
  defp term(s) do
    OK.for do
      {x,s}<-factor(s)
      res<-term(s,x)
    after
      res
    end
  end
  defp term("*"<>s,n) do
    OK.for do
      {x,s}<-factor(s)
      res<-term(s,Rational.mul(n,x))
    after
      res
    end
  end
  defp term("/"<>s,n) do
    OK.for do
      {x,s}<-factor(s)
      res<-term(s,Rational.div(n,x))
    after
      res
    end
  end
  defp term(s,n),do: {:ok,{n,s}}
  defp factor("("<>s) do
    OK.for do
      {x,s}<-expr(s)
      {_,s}<-char(s,")")
    after
      {x,s}
    end
  end
  defp factor(s),do: number(s)
end

File: ts/parser.ts
import * as Rational from "./rational";

export class Parser {
    pos = 0;
    constructor(public s: string) {
    }

    peek() {
        if (this.pos >= this.s.length) {
            throw new Error();
        }
        return this.s[this.pos];
    }

    next() {
        let v = this.peek();
        this.pos++;
        return v;
    }

    expect(f: (x: string) => boolean) {
        let v = this.peek();
        if (f(v)) {
            this.next();
            return v;
        } else {
            throw new Error();
        }
    }

    char(c: string) {
        return this.expect(v => v === c);
    }

    number() {
        let g;
        let s = "";
        try {
            this.char("-");
            g = -1;
        } catch{
            g = 1;
        }

        while (true) {
            try {
                s += this.expect(isDigit);
            } catch{
                if (s.length === 0) {
                    throw new Error();
                } else {
                    break;
                }
            }
        }

        if (s.length !== 1 && s[0] === "0") {
            throw new Error();
        }
        const v = +s * g;
        if (isNaN(v)) {
            throw new Error();
        } else {
            return Rational.rational(v, 1);
        }
    }

    eof() {
        if (this.s.length !== this.pos) {
            throw new Error();
        }
    }

    factor(): Rational.Rational {
        try {
            this.char("(");
        } catch{
            return this.number();
        }

        let v = this.expr();
        this.char(")");
        return v;
    }

    term(): Rational.Rational {
        let x = this.factor();
        while (true) {
            let op;
            try {
                op = this.expect(c => c === "*" || c === "/");
            } catch{
                break;
            }

            switch (op) {
                case "*":
                    x = Rational.mul(x, this.factor());
                    break;
                case "/":
                    x = Rational.div(x, this.factor());
                    break;
            }
        }
        return x;
    }

    expr(): Rational.Rational {
        let x = this.term();
        while (true) {
            let op;
            try {
                op = this.expect(c => c === "+" || c === "-");
            } catch{
                break;
            }

            switch (op) {
                case "+":
                    x = Rational.add(x, this.term());
                    break;
                case "-":
                    x = Rational.sub(x, this.term());
                    break;
            }
        }
        return x;
    }

    parse() {
        let v = this.expr();
        this.eof();
        return v;
    }
}

function isDigit(c: string) {
    return "0123456789".includes(c);
}